fix: Make value required on AriaCheckboxGroupItemProps interface

stefanprobst opened the pull request:
Closes  #1318

Makes `value` required on `AriaCheckboxGroupItemProps`. The doc comment for `value` is inherited from `ToggleProps`.

## ✅ Pull Request Checklist:

- [x] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

<!--- Include instructions to test this pull request -->

## 🧢 Your Project:

<!--- Company/project for pull request -->

stefanprobst commented:
hmm, this would require changes to the `SpectrumCheckboxProps` as well, otherwise ts will complain about [this](https://github.com/adobe/react-spectrum/blob/2cacc4f9cde659f7295a333bbfa8d5e052b0460d/packages/%40react-spectrum/checkbox/src/Checkbox.tsx#L52).

snowystinger commented:
Hey! Thanks for working on this one. There are two things we can do here to help things out.
1) Fix the aria hook types. You've mostly done this, I think we can either
- @ts-ignore
- or just add value: props.value explicitly rather than relying on the …props
in order to resolve the TS error, just make sure to add a comment explaining why we're ignoring it

2) Help React Spectrum users.
There are two different types and the shift in which one is used is based on the React tree when rendered
value not required
```
<Checkbox />
```
vs
value is required
```
<CheckboxGroup>
  <div>
    <span>
     <...whatever HoC/wrapper >
       <Checkbox value="i-use-context" />
```

Typescript won't know the tree during compilation, this is only at run time.
I think we're going to have to solve this by throwing an error in `useCheckboxGroupItem` if there is no props.value

stefanprobst commented:
thanks for reviewing!

added `value: props.value` (and a note that this works as long as `strictNullChecks` is disabled).

and added a `console.warn` in case of missing `value` on a checkboxgroup item, since other disallowed props (e.g. `isSelected`) are also handled that way. if you think it's better to throw, i'll adjust.

sidenote: any reason to not make `Checkbox` and `CheckboxGroupItem` two components on top of a `CheckboxBase` (like with `TextFieldBase` or `ListBoxBase`? then both could be explicit about which props they accept (and also avoid the conditional hook call?)

snowystinger commented:
I had the thought to create a second component, but it would be a breaking change to pull that out of Checkbox as it currently is. It's also more to educate people on that a Checkbox can be used alone, but in a group they must use this specialized Checkbox. It might be something to explore further down the road.

devongovett commented:
Mainly we wanted to avoid unnecessarily large API surface area. IMO it seems a bit much to make a whole new component just to make one prop required. 🤷

devongovett commented on packages/@react-spectrum/checkbox/src/Checkbox.tsx:
```diff
@@ -68,6 +71,9 @@ function Checkbox(props: SpectrumCheckboxProps, ref: FocusableRef<HTMLLabelEleme
         console.warn(`${key} is unsupported on individual <Checkbox> elements within a <CheckboxGroup>. Please apply these props to the group instead.`);
       }
     }
+    if (!props.value) {
```
Do an explicit null check in case the value is an empty string or something

```suggestion
    if (props.value == null) {
```

stefanprobst commented:
> Mainly we wanted to avoid unnecessarily large API surface area. IMO it seems a bit much to make a whole new component just to make one prop required.

makes sense! (it's [`isSelected`, `defaultSelected`, `isEmphasized`](https://github.com/adobe/react-spectrum/blob/1d9905e7c8308cfbb2a1e03bfc22b2dce89cbb12/packages/%40react-spectrum/checkbox/src/Checkbox.tsx#L66) props as well however)

snowystinger reviewed (approved):
Looks good to me!

ktabors reviewed (approved):
LGTM
